design/Extend-VolumePolicies-to-support-more-actions.md

```diff
@@ -0,0 +1,176 @@
+# Extend VolumePolicies to support more actions
+
+## Abstract
+
+Currently, the [VolumePolicies feature](https://github.com/vmware-tanzu/velero/blob/main/design/Implemented/handle-backup-of-volumes-by-resources-filters.md) which can be used to filter/handle volumes during backup only supports the skip action on matching conditions. Users need more actions to be supported.
+
+## Background
+
+The `VolumePolicies` feature was introduced in Velero 1.11 as a flexible way to handle volumes. The main agenda of
+introducing the VolumePolicies feature was to improve the overall user experience when performing backup operations
+for volume resources, the feature enables users to group volumes according the `conditions` (criteria) specified and
+also lets you specify the `action` that velero needs to take for these grouped volumes during the backup operation.
+The limitation being that currently `VolumePolicies` only supports `skip` as an action, We want to extend the `action`
+functionality to support more usable options like `fs-backup` (File system backup) and `snapshot` (VolumeSnapshots).
+
+## Goals
+- Extending the VolumePolicies to support more actions like `fs-backup` (File system backup) and `snapshot` (VolumeSnapshots).
+- Improve user experience when backing up Volumes via Velero
+
+## Non-Goals
+- No changes to existing approaches to opt-in/opt-out annotations for volumes
+- No changes to existing `VolumePolicies` functionalities
+
+
+## Use-cases/Scenarios
+
+**Use-case 1:**
+- A user wants to use `snapshot` (volumesnapshots) backup option for all the csi supported volumes and `fs-backup` for the rest of the volumes.
```

Comment:
@shawn-hurley I think that would be out-of-scope for this design. The task here is to provide an alternate means of telling Velero which volumes to use fs backup vs. which volumes to snapshot. We don't yet support "some datamover, others snapshot-without-datamover", much less "some this datamover, others that datamover" -- those selections are once-per-backup rather than once-per-volume, but in the future, it's possible that we could add volume-level policies for those new action types.

Reply:
In the initial design, there's a `Parameters` field under each `Action` struct, so `datamover` and other params can be set as parameters of the `snapshot` action type.

cc @qiuming-best

Reply:
Extending the snapshot selection per backup with the parameters to define which data mover would be an excellent addition. This allows users to start at least building data movers.

Reply:
At present, we have the `--data-mover `parameter per backup, not sure if we need to make it per volume. Probably, we can keep the current behavior until we see the concrete requirement for the per volume one.

Reply:
I honestly don't think that anyone will build one until you can do it by volume.

Reply:
@shawn-hurley I agree that we want to consider this, but I still think it's out-of-scope for this particular design/implementation. I'm wary of scope creep here since we want to get this implemented in 1.14. The goal here is to use volume policies to replace the behavior currently managed via per-volume annotations to determine when to use fs-backup and when to snapshot. Following on from this we could propose an additional design to manage two additional things via volume policies:
1) on a per-volume/per-policy basis, use datamover (currently "use datamover vs . just CSI snapshots" is per-backup)
2) on a per-volume/per-policy basis, choose which datamover (this is also currently per-backup).